### ISSUE TITLE
Getting "measure" tool working again.

### ISSUE THREAD

**Issue by TrevorHansen** [stp/stp#280]
Currently I'm trying to fix the "measure" tool. The measure tool is helpful because we can use it to measure how well different encodings of basic operations propagate. So given two different encodings of signed-less than (say), if one encoding propagates better (that is, the SAT solver deduces the assignment to more bits) and if it is also converted into a smaller number of CNF clauses, then we consider that encoding to be better than the other one.

To get the measure tool working I need the SAT solver (with assumptions) just to do unit propagation, with not searching.

Previously, I'd hacked minisat to get it working, but that code has been removed since we moved to using the upstream minisat. Does anyone know how to do it with Minisat, or if instead we can use Cryptominisat to do it?

Here's the current (broken) code from MinisatCore.cpp
```
// *Doesn't solve*, just does a single unit propagate.
// returns false if UNSAT.
bool MinisatCore::propagateWithAssumptions(
    const stp::SATSolver::vec_literals& assumps)
{
  if (!s->simplify())
    return false;

  setMaxConflicts(0);
  Minisat::lbool ret = s->solveLimited(assumps);
  assert(s->conflicts ==0);
  return ret != (Minisat::lbool)l_False;
}
```
A complication is that because we mostly don't have a model, we can't read the modelValue, so try reading the value. It doesn't seem to work though.

Trev.

**Comment by msoos** [comment]
OK, it works as I expected. I added some tests to make this clear. Note that `str_to_cl` is just a helper and "1" in there is `Lit(0, false)` and "-2"  is `Lit(1, true)`.

Here is a simple test:
```
TEST(propagate, prop_1)
{
    SATSolver s;
    s.new_vars(10);
    s.add_clause(str_to_cl("1, 2"));
    s.add_clause(str_to_cl("-1"));

    vector<Lit> lits = s.get_zero_assigned_lits();
    vector<Lit>::iterator it;
    it = std::find(lits.begin(), lits.end(), Lit(0, true));
    EXPECT_NE(it, lits.end());
    it = std::find(lits.begin(), lits.end(), Lit(1, false));
    EXPECT_NE(it, lits.end());

    EXPECT_EQ(lits.size(), 2);
}
```
So both "1" and "2" have been set (internally, 0, and 1).

Here is a more complex one:

```
TEST(propagate, prop_complex)
{
    SATSolver s;
    s.new_vars(30);

    s.add_clause(str_to_cl("-1"));
    s.add_clause(str_to_cl("1, 2"));

    s.add_clause(str_to_cl("-5, 6"));
    s.add_clause(str_to_cl("5"));

    s.add_clause(str_to_cl("1, -2, -5, -6, 7"));

    vector<Lit> lits = s.get_zero_assigned_lits();
    vector<Lit>::iterator it;
    it = std::find(lits.begin(), lits.end(), Lit(6, true));
    EXPECT_EQ(lits.size(), 5);
}
```

So this propagates everything, as you can see. You can query what has been propagated with `get_zero_assigned_lits()` and it will give you a list of complete and non-repeating literals that have been propagated.

If this fixes the issue, please close :) Otherwise, let me know what I misunderstood :) Thanks!

**Comment by TrevorHansen** [comment]
Thanks Mate!

I've checked in a first attempt as below. It seems like the way I've done it the assumptions aren't being used to deduce level0 assignments? Can it be done using assumptions, or should I send all the clauses each time?


```
// Count how many literals/bits get fixed subject to the assumptions..
uint32_t CryptoMiniSat5::getFixedCountWithAssumptions(const stp::SATSolver::vec_literals& assumps)
{
  std::cerr << s->get_zero_assigned_lits().size() << " assignments at start" <<std::endl;

  //assert(s->get_zero_assigned_lits().size() == 0);
  const uint64_t conf = s->get_sum_conflicts();
  assert(conf == 0);
  
  vector<CMSat::Lit>& real_temp_cl = *(vector<CMSat::Lit>*)temp_cl;
  real_temp_cl.clear();
  for (int i = 0; i < assumps.size(); i++)
  {
    real_temp_cl.push_back(CMSat::Lit(var(assumps[i]), sign(assumps[i])));
  }

  std::cerr << real_temp_cl.size() << " assumptions" << std::endl;

  const CMSat::lbool r = s->simplify(&real_temp_cl);

  std::cerr << s->get_zero_assigned_lits().size() << " assignments at end" <<std::endl;

  // The assumptions are each single literals (corresponding to bits) that are true/false. 
  // so in the result they should be all be set 
  assert(s->get_zero_assigned_lits().size() >= assumps.size());
  assert(s->get_sum_conflicts() == conf ); // no searching, so no conflicts.
  assert(CMSat::l_False != r); // always satisfiable.

  return s->get_zero_assigned_lits().size();
}
```

**Comment by msoos** [comment]
Yep, assumptions are not being used to deduce zero-level assignments. In case you need how many variables would be propagated given the assumptions, I can create a function for that. It's very easy for me. Otherwise, you'd need to re-initialise all the clauses all the time, which may be too slow. I mean, I don't understand your use case completely -- in case you'd put the clauses in all the time anyway, because they are always different, then just create a new Solver() every time and it's done. But if you want to do:

```
Solver S;
//add a million clauses to S
for (int i =0; i <1000; i++) {
 S.add_clause([clause with Lit(i, true)]);
 ....
 S.add_clause([clause with Lit(i, true)]);
 vector<Lit> assums;
 assums.push_back(Lit(i, false));
 int num = S.get_assigned_lits_given_assumptions(assums).size();
 cout << "It's " << num << endl;
}
```
Then of course you'd be suffering a lot because you'd need to add those million clauses every time to a fresh Solver. How many clauses are we talking about? What is the max `i` ?

Also, beware, because simplify() will do a **lot** of things, including a form of of conflict generation, and a bunch of other stuff. It's also not cheap.

**Comment by TrevorHansen** [comment]
Thanks Mate,

I'll try sending all the clauses each time and see if it's fast enough. There's a small number of clauses (<65K), but in the past i've run each operation with hundreds of thousands of random partial assignments. 

The goal is to select the best encoding for basic operations. So i'd like to simulate as closely as possible how strongly CMS will propagate on different encodings of the same operation. It sounds like it's best to:
1) Send the encoding to CMS
2) run simplify()
3) Send the partial assignment as clauses to CMS.
4) Read the literals assigned at level-0

Then given two encodings of the same operation if one of the encodings has fewer clauses and across 100K random partial assignments deduces more input or output bits/literals, then it's the better of the two encodings, so we should use it in STP.

Is this how you would do it?

**Comment by msoos** [comment]
The above works, yes. Note that if you are not using assumptions (i.e. inserting the clauses all the time) then you are probably best off running simplify() **after** you added the partial assignements. It better simulates what CMS can do once the assignements are there. The simplify() will work better then. In case you will use the assumptions, then the above is probably best, but please note that simplify() actually takes assumptions, and it is *highly recommended* that you put your future assumptions in there. Otherwise, CMS will need to do some funky stuff to undo the simplifications on the assumed variables, and that is just not good practice and may not help speed.

So in case you don't use assumptions (and need my help implementing the thing I proposed above) then invert steps 2 and 3. Otherwise, keep the steps, but add the assumptions into 2.

**Comment by TrevorHansen** [comment]
Thanks Mate, 

I've checked in a version without assumptions that seems to work (5c0e9a0baa757f333b945238370ede2a9036172f).

It's about 100 times slower than with assumptions, but I think it's good enough. I'll check whether it's working properly in the next few days.

**Comment by msoos** [comment]
OK, thanks! Note that I can do the assumption stuff in about 20-30 minutes so, don't suffer for that :) 

Note that if in the real world you'll be adding the unit clauses instead of doing assumptions, then the way you are doing it now is better. The simplifications don't work with assumed variables (they cannot, or it would need to save the solver state, etc).

**Comment by TrevorHansen** [comment]
Using the "measure" tool, I've decided to change (8777c6349e3aac9d945a0298c69af4fc0b7d14e3)  the comparison encoding. The new encoding is smaller and propagates as strongly as the old one.

It makes me wonder whether a better way to compare two encodings, for example comparison is:
1) To randomly generate a satisfying assignment,  e.g. ( 0x08 < 0x10 ) is false.
2) Then to randomly select a bit, which is unknown at level 0, to assert to the SAT solver, until all the bits are determined at level 0.

Across some number of samples (e.g. 10k ) the encoding that assigns all the bits at level 0, given the fewest number of bits asserted, is the encoding which propagates best.

Is there a better way?

**Comment by TrevorHansen** [comment]
ooops. ( 0x08 < 0x10 ) is true.

**Comment by msoos** [comment]
This way seems OK. Note that some variables are better to be asserted -- so check what variables get asserted -- if only some intermediary ones are asserted, maybe that's not so good (you may want stronger propagation properties). So just counting the number of variables asserted at level 0 may not be enough, as some may be better to be asserted than others. Similarly, not all toplevel variables are the same -- some may propagate more than others, but here you are compensating by doing e.g. 10k trials. So I think that's OK and you don't need to compensate for that (just do a lot of randomised trials). But what variables get propagated you may want to distinguish -- e.g. by giving 2x the "points" for a variable that is important from the SMT perspective.

BELOW IS OFF-TOPIC
/////
Note that we want to eventually push back zero-level assignements to STP. So that after a `solve()` we could get top level assignments, push them into STP, then propagate at the STP level. In fact, we could  do a `solve()` with a short timeout, check if anything gets set at toplevel, push that into STP, propagate at the STP/theory level, push newly propagated variables into CMS, let CMS run again for a while, etc. This setup I think could beat a lot of SMT solvers out there...
